### ISSUE TITLE
False positive unnecessary-comprehension

### ISSUE THREAD

**Issue by sam-s** [PyCQA/pylint#3164]
### Steps to reproduce
```python
def g(x,y,z): return x+y+z
d = {}
b = {k:v for k,v in [(k,g(1,x,y)) for k,(x,y) in d.items()] if v}

print(["%2d %15s %d  %s" % (i,f,-n,l)
       for i,(n,f,l) in enumerate(sorted([
               (-len(l),f,l) for f,l in d.items()]))])
```

### Current behavior

`(R1721, unnecessary-comprehension, ) Unnecessary use of a comprehension` 
on lines `b = ...` and the last line (`print`).

Note that `b=` does **two** comprehensions because `g` is a very expensive function so I want to call it only once, and I want to filter the results, so one of the comprehensions filters.

If there is a better way to express what I am trying to do, I am open to suggestions.

Thanks!

### Expected behavior

no diagnistics

### pylint --version output
```
pylint 2.4.2
astroid 2.3.1
Python 3.7.4 (default, Sep  7 2019, 18:27:02) 
[Clang 10.0.1 (clang-1001.0.46.4)]
```

**Comment by PCManticore** [comment]
Thanks for reporting the issue. This is already fixed in the `master` branch with https://github.com/PyCQA/pylint/commit/05d3e15d9f6978b037943625fe09b9a384ea63c2 and it will be part of `2.4.3` relatively soon (this week).

**Comment by jdhao** [comment]
@PCManticore  I am seeing the unnecessary-comprehension warning for the following code:

```python
labels = ['one', 'two', 'three']                             
indexed_labels = [(i, text) for i, text in enumerate(labels)]
```

I am current using pylint 2.4.4. I think it is a false positive.
